osu.Game/Overlays/News/Sidebar/YearsPanel.cs

```diff
@@ -0,0 +1,165 @@
+            {
+                IdleColour = colourProvider.Light2;
+                HoverColour = colourProvider.Light1;
+                Action = () => { }; // Avoid button being disabled since there's no proper action assigned.
+            }
+        }
+
+        private class YearsGridContainer : GridContainer
+        {
+            private const int column_count = 4;
+            private const float spacing = 5f;
+
+            private readonly int rowCount;
+            private readonly int[] years;
+
+            public YearsGridContainer(int[] years)
+            {
+                this.years = years;
+                rowCount = (int)Math.Ceiling((float)years.Length / column_count);
+
+                RelativeSizeAxes = Axes.X;
+                AutoSizeAxes = Axes.Y;
+                RowDimensions = getRowDimensions();
+                ColumnDimensions = getColumnDimensions();
+                Content = createContent();
+            }
+
+            private Dimension[] getRowDimensions()
+            {
+                var rowDimensions = new Dimension[rowCount];
+                for (int i = 0; i < rowCount; i++)
+                    rowDimensions[i] = new Dimension(GridSizeMode.AutoSize);
+
+                return rowDimensions;
+            }
+
+            private Dimension[] getColumnDimensions()
+            {
+                var columnDimensions = new Dimension[column_count];
+                for (int i = 0; i < column_count; i++)
+                    columnDimensions[i] = new Dimension(GridSizeMode.Relative, size: 1f / column_count);
+
+                return columnDimensions;
+            }
+
+            private Drawable[][] createContent()
+            {
+                var buttons = new Drawable[rowCount][];
+
+                for (int i = 0; i < rowCount; i++)
+                {
+                    buttons[i] = new Drawable[column_count];
+
+                    for (int j = 0; j < column_count; j++)
+                    {
+                        var index = i * column_count + j;
+                        buttons[i][j] = index >= years.Length
+                            ? Empty()
+                            : new Container
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                AutoSizeAxes = Axes.Y,
+                                Padding = new MarginPadding
+                                {
+                                    Top = i == 0 ? 0 : spacing / 2,
+                                    Bottom = i == rowCount - 1 ? 0 : spacing / 2,
+                                    Left = j == 0 ? 0 : spacing / 2,
+                                    Right = j == column_count - 1 ? 0 : spacing / 2
+                                },
+                                Child = new YearButton(years[index])
+                            };
+                    }
+                }
+
+                return buttons;
+            }
```

Comment:
FWIW this could be done in a *slightly* cleaner way by attaching empty cells between each year instead:

<details>
<summary>changes</summary>

```diff
diff --git a/osu.Game/Overlays/News/Sidebar/YearsPanel.cs b/osu.Game/Overlays/News/Sidebar/YearsPanel.cs
index 2a4e2024d2..3c24e580b8 100644
--- a/osu.Game/Overlays/News/Sidebar/YearsPanel.cs
+++ b/osu.Game/Overlays/News/Sidebar/YearsPanel.cs
@@ -96,6 +96,9 @@ private class YearsGridContainer : GridContainer
             private const int column_count = 4;
             private const float spacing = 5f;

+            private Dimension[] rowDimensions;
+            private Dimension[] columnDimensions;
+
             private readonly int rowCount;
             private readonly int[] years;

@@ -106,55 +109,67 @@ public YearsGridContainer(int[] years)

                 RelativeSizeAxes = Axes.X;
                 AutoSizeAxes = Axes.Y;
-                RowDimensions = getRowDimensions();
-                ColumnDimensions = getColumnDimensions();
+                RowDimensions = rowDimensions = getRowDimensions();
+                ColumnDimensions = columnDimensions = getColumnDimensions();
                 Content = createContent();
             }

             private Dimension[] getRowDimensions()
             {
-                var rowDimensions = new Dimension[rowCount];
+                var dimensions = new List<Dimension>();
+
                 for (int i = 0; i < rowCount; i++)
-                    rowDimensions[i] = new Dimension(GridSizeMode.AutoSize);
+                {
+                    dimensions.Add(new Dimension(GridSizeMode.AutoSize));

-                return rowDimensions;
+                    if (i != rowCount - 1)
+                        dimensions.Add(new Dimension(GridSizeMode.Absolute, spacing));
+                }
+
+                return dimensions.ToArray();
             }

             private Dimension[] getColumnDimensions()
             {
-                var columnDimensions = new Dimension[column_count];
+                var dimensions = new List<Dimension>();
+
                 for (int i = 0; i < column_count; i++)
-                    columnDimensions[i] = new Dimension(GridSizeMode.Relative, size: 1f / column_count);
+                {
+                    dimensions.Add(new Dimension(GridSizeMode.Distributed));

-                return columnDimensions;
+                    if (i != column_count - 1)
+                        dimensions.Add(new Dimension(GridSizeMode.Absolute, spacing));
+                }
+
+                return dimensions.ToArray();
             }

             private Drawable[][] createContent()
             {
-                var buttons = new Drawable[rowCount][];
+                var buttons = new Drawable[rowDimensions.Length][];

-                for (int i = 0; i < rowCount; i++)
-                {
-                    buttons[i] = new Drawable[column_count];
+                int y = 0;

-                    for (int j = 0; j < column_count; j++)
+                for (int i = 0; i < rowDimensions.Length; i++)
+                {
+                    if (rowDimensions[i].Mode == GridSizeMode.Absolute)
+                        buttons[i] = new[] { Empty() };
+                    else
                     {
-                        var index = i * column_count + j;
-                        buttons[i][j] = index >= years.Length
-                            ? Empty()
-                            : new Container
+                        buttons[i] = new Drawable[columnDimensions.Length];
+
+                        for (int j = 0; j < columnDimensions.Length; j++)
+                        {
+                            if (columnDimensions[j].Mode == GridSizeMode.Absolute)
+                                buttons[i][j] = Empty();
+                            else
                             {
-                                RelativeSizeAxes = Axes.X,
-                                AutoSizeAxes = Axes.Y,
-                                Padding = new MarginPadding
-                                {
-                                    Top = i == 0 ? 0 : spacing / 2,
-                                    Bottom = i == rowCount - 1 ? 0 : spacing / 2,
-                                    Left = j == 0 ? 0 : spacing / 2,
-                                    Right = j == column_count - 1 ? 0 : spacing / 2
-                                },
-                                Child = new YearButton(years[index])
-                            };
+                                if (y >= years.Length)
+                                    buttons[i][j] = Empty();
+                                else
+                                    buttons[i][j] = new YearButton(years[y++]);
+                            }
+                        }
                     }
                 }
```

</details>

But either way this sounds like something that should be supported framework-side as "grid column / row gaps", later on when more usages of this pop up.

Reply:
To be honest, your logic seems way too hacky (storing dimensions and then comparing their modes), but that's just my preference at this point.

Reply:
Hmm, to be fair if this was to be implemented in `GridContainer` it would probably go the way you have it, so yeah disregard that, I think this is good for now.